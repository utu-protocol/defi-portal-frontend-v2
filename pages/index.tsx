import WalletConnectProvider from '@walletconnect/web3-provider'
import { providers } from 'ethers'
import Head from 'next/head'
import { useCallback, useEffect, useReducer } from 'react'
// import WalletLink from 'walletlink'
import Web3Modal from 'web3modal'
import { getChainData } from '../lib/utilities'

import Navigation from "../components/Navigation"
import Table from "../components/Table"

const INFURA_ID = '460f40a260564ac4a4f4b3fffb032dad'

const providerOptions = {
  walletconnect: {
    package: WalletConnectProvider, // required
    options: {
      infuraId: INFURA_ID, // required
    },
  },
  // 'custom-walletlink': {
  //   display: {
  //     logo: 'https://play-lh.googleusercontent.com/PjoJoG27miSglVBXoXrxBSLveV6e3EeBPpNY55aiUUBM9Q1RCETKCOqdOkX2ZydqVf0',
  //     name: 'Coinbase',
  //     description: 'Connect to Coinbase Wallet (not Coinbase App)',
  //   },
  //   options: {
  //     appName: 'Coinbase', // Your app name
  //     networkUrl: `https://mainnet.infura.io/v3/${INFURA_ID}`,
  //     chainId: 1,
  //   },
  //   package: WalletLink,
  //   connector: async (_, options) => {
  //     const { appName, networkUrl, chainId } = options
  //     const walletLink = new WalletLink({
  //       appName,
  //     })
  //     const provider = walletLink.makeWeb3Provider(networkUrl, chainId)
  //     await provider.enable()
  //     return provider
  //   },
  // },
}

let web3Modal
if (typeof window !== 'undefined') {
  web3Modal = new Web3Modal({
    network: 'mainnet', // optional
    cacheProvider: true,
    providerOptions, // required
  })
}

type StateType = {
  provider?: any
  web3Provider?: any
  address?: string
  chainId?: number
}

type ActionType =
  | {
      type: 'SET_WEB3_PROVIDER'
      provider?: StateType['provider']
      web3Provider?: StateType['web3Provider']
      address?: StateType['address']
      chainId?: StateType['chainId']
    }
  | {
      type: 'SET_ADDRESS'
      address?: StateType['address']
    }
  | {
      type: 'SET_CHAIN_ID'
      chainId?: StateType['chainId']
    }
  | {
      type: 'RESET_WEB3_PROVIDER'
    }

const initialState: StateType = {
  provider: null,
  web3Provider: null,
  address: null,
  chainId: null,
}

function reducer(state: StateType, action: ActionType): StateType {
  switch (action.type) {
    case 'SET_WEB3_PROVIDER':
      return {
        ...state,
        provider: action.provider,
        web3Provider: action.web3Provider,
        address: action.address,
        chainId: action.chainId,
      }
    case 'SET_ADDRESS':
      return {
        ...state,
        address: action.address,
      }
    case 'SET_CHAIN_ID':
      return {
        ...state,
        chainId: action.chainId,
      }
    case 'RESET_WEB3_PROVIDER':
      return initialState
    default:
      throw new Error()
  }
}

export const Home = (): JSX.Element => {
  const [state, dispatch] = useReducer(reducer, initialState)
  const { provider, web3Provider, address, chainId } = state

  const connect = useCallback(async function () {
    // This is the initial `provider` that is returned when
    // using web3Modal to connect. Can be MetaMask or WalletConnect.
    const provider = await web3Modal.connect()

    // We plug the initial `provider` into ethers.js and get back
    // a Web3Provider. This will add on methods from ethers.js and
    // event listeners such as `.on()` will be different.
    const web3Provider = new providers.Web3Provider(provider)

    const signer = web3Provider.getSigner()
    const address = await signer.getAddress()

    const network = await web3Provider.getNetwork()

    dispatch({
      type: 'SET_WEB3_PROVIDER',
      provider,
      web3Provider,
      address,
      chainId: network.chainId,
    })
  }, [])

  const disconnect = useCallback(
    async function () {
      await web3Modal.clearCachedProvider()
      if (provider?.disconnect && typeof provider.disconnect === 'function') {
        await provider.disconnect()
      }
      dispatch({
        type: 'RESET_WEB3_PROVIDER',
      })
    },
    [provider]
  )

  // Auto connect to the cached provider
  useEffect(() => {
    if (web3Modal.cachedProvider) {
      connect()
    }
  }, [connect])

  // A `provider` should come with EIP-1193 events. We'll listen for those events
  // here so that when a user switches accounts or networks, we can update the
  // local React state with that new information.
  useEffect(() => {
    if (provider?.on) {
      const handleAccountsChanged = (accounts: string[]) => {
        // eslint-disable-next-line no-console
        console.log('accountsChanged', accounts)
        dispatch({
          type: 'SET_ADDRESS',
          address: accounts[0],
        })
      }

      const handleChainChanged = (accounts: string[]) => {
        // eslint-disable-next-line no-console
        console.log('accountsChanged', accounts)
        dispatch({
          type: 'SET_ADDRESS',
          address: accounts[0],
        })
      }

      const handleDisconnect = (error: { code: number; message: string }) => {
        // eslint-disable-next-line no-console
        console.log('disconnect', error)
        disconnect()
      }

      provider.on('accountsChanged', handleAccountsChanged)
      provider.on('chainChanged', handleChainChanged)
      provider.on('disconnect', handleDisconnect)

      // Subscription Cleanup
      return () => {
        if (provider.removeListener) {
          provider.removeListener('accountsChanged', handleAccountsChanged)
          provider.removeListener('chainChanged', handleChainChanged)
          provider.removeListener('disconnect', handleDisconnect)
        }
      }
    }
  }, [provider, disconnect])

  const chainData = getChainData(chainId)

  return (
    <div className="bg-primary h-80">
      <Head>
        <title>Ocean Market by UTU</title>
        <link rel="icon" href="/favicon.ico" />
      </Head>
    <div>
    <Navigation connect={connect} disconnect={disconnect} web3Provider={web3Provider} address={address} chainData={chainData}/>      <h1 className="max-w-7xl px-8 py-12  mx-auto text-3xl font-bold text-white">
        Ocean Market
      </h1>
    </div>
    <div className="-mt-8">
      <Table />
    </div>
    </div>
  )
}

export default Home
