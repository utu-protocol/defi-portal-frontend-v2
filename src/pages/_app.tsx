import { ReactElement } from 'react'
import type { AppProps, NextWebVitalsMetric } from 'next/app'
import '../styles/globals.css'

function MyApp({ Component, pageProps }: AppProps): ReactElement {
  return <Component {...pageProps} />
}

export function reportWebVitals(metric: NextWebVitalsMetric): void {
  // eslint-disable-next-line no-console
  console.log(metric)
}

export default MyApp
