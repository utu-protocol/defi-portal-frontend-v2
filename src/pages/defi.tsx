import { ReactElement, useEffect, useState } from 'react'

import Table from '../components/Table'
import Layout from '../components/Layout'
import { useSelector } from 'react-redux'

export const Home = (): ReactElement => {
  const [list, setList] = useState([])
  const { address } = useSelector((state: any) => ({
    address: state.wallet.address,
  }))
  const fetchData = async () => {
    const data = {
      sourceCriteria: JSON.stringify({
        type: 'Address',
        ids: { address },
      }),
      targetType: 'DefiPortal',
    }
    const res = await fetch(
      `${
        process.env.REACT_APP_UTU_API_BASE_URL
      }/core-api/ranking?${new URLSearchParams(
        Object.entries(data)
      ).toString()}`
    )
    const { result } = await res.json()
    setList(result)
  }

  useEffect(() => {
    fetchData()
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
  return (
    <Layout title="DeFi">
      <Table list={list} />
    </Layout>
  )
}

export default Home
