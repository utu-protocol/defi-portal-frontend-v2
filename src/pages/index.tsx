import { ReactElement } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import Layout from '../components/Layout'

import { connect } from '../redux/actions/wallet.actions'

export const Home = (): ReactElement => {
  const dispatch = useDispatch()
  const { address } = useSelector((state: any) => state.wallet)
  // const [ethAddress, setEthAddress] = useState('')
  // const [errorMsg, setErrorMsg] = useState('')

  // const openDashboard = async () => {
  //   const result = await dispatch(storeAddress({ address: ethAddress }))
  //   if (!result) {
  //     setErrorMsg('Please provide a valid Ethereum address')
  //     return
  //   }
  // }

  // const onChangeAddress = (e: any) => {
  //   if (errorMsg) {
  //     setErrorMsg('')
  //   }

  //   setEthAddress(e.target.value)
  // }


  return (
    <Layout
    // title="Tailored DeFi recommendations"
    // subtitle="From the people you trust."
    >
      <div className="max-w-screen-lg mx-5 pt-44 mt-2">
        <h2 className="text-3xl leading-9 font-extrabold tracking-tight text-white sm:text-4xl sm:leading-10">
          Tailored DeFi recommendations
          <br />
          <span id="newsletter-headline" className="text-portal">
            From the people <span className="underline">you</span> trust.
          </span>
        </h2>
        <p className="mt-4 pt-5 text-lg leading-7 text-gray-500 max-w-sm">
          Get personalized DeFi recommendations from the people you trust.
          Leveraging the power of the UTU Trust Engine, you can now navigate the
          Decentralized Finance world safely.
        </p>
        {!address && (
          <>
            <div className="mt-8 sm:flex" aria-labelledby="newsletter-headline">
              {/* <div className="w-full sm:max-w-lg">
                <input
                  onChange={onChangeAddress}
                  value={ethAddress}
                  aria-label="address"
                  type="text"
                  className="font-mono appearance-none w-full px-5 py-3 border border-gray-300 text-base leading-6 rounded-md text-gray-900 bg-white placeholder-gray-500 focus:outline-none focus:shadow-outline focus:border-blue-300 transition duration-150 ease-in-out"
                  placeholder="Enter ENS domain or a Ethereum address"
                />
              </div> */}
              {/* <div className="flex mt-3 rounded-md shadow sm:mt-0 sm:ml-3 sm:flex-shrink-0">
                <button
                  onClick={openDashboard}
                  className="w-full flex items-center justify-center px-5 py-3 border border-transparent text-base leading-6 font-medium rounded-md text-white bg-black hover:shadow-md focus:outline-none transition duration-150 ease-in-out"
                >
                  {"Let's Go!"}
                </button>
              </div>
              <p className="ml-4 mr-2 mt-3">or</p> */}
              <div className="flex mt-3 rounded-md shadow sm:mt-0 sm:flex-shrink-0">
                <button
                  type="button"
                  onClick={() => {
                    dispatch(connect())
                  }}
                  className="bg-primary w-full flex items-center justify-center px-5 py-3 border border-transparent text-base leading-6 font-medium rounded-md text-white hover:shadow-md focus:outline-none transition duration-150 ease-in-out"
                >
                  Connect Wallet
                </button>
              </div>
            </div>
            {/* <p className="text-red-500 text-sm mt-1">{errorMsg && errorMsg}</p> */}
          </>
        )}
      </div>
    </Layout>
  )
}

export default Home
