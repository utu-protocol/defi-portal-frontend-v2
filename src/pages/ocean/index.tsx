/* eslint-disable react-hooks/exhaustive-deps */
import { ReactElement, useEffect, useMemo } from 'react'

import Table from '../../components/Table';
import Layout from '../../components/Layout'
import { useSelector } from 'react-redux'
import { useDispatch } from 'react-redux'
import { fetchAssets } from '../../redux/actions/ocean.actions'
import Loader from '../../components/Loader'
import { Link } from 'react-router-dom';

export const Ocean = (): ReactElement => {
  const dispatch = useDispatch()
  const { assets, loading } = useSelector((state: any) => ({
    assets: state.ocean.assets,
    loading: state.ocean.loading,
  }))
  useEffect(() => {
    dispatch(fetchAssets())
  }, []);


  const data = useMemo(
    () => {
      return assets.map(({ entity, summaryText }: any) => ({
        col1: {
          name: entity.name,
          address: entity.ids.address_datatoken,
          publisher: entity.properties.PublishedBy
        },
        col2: entity.properties.Purgatory ? 1 : 0,
        col3: entity.properties.PublishedBy,
        col4: summaryText,
        col5: entity.properties.Consumed,
      }))
    },
    [assets]
  );

  const columns = useMemo(
    () => [
      {
        Header: 'Name',
        id: 'name',
        accessor: 'col1',
        Cell: ({ value }: any) => {
          return (<Link to={`/ocean-market/${value.address}`}>
            <div className="text-sm font-medium text-gray-900">{value.name}</div>
            <div className="text-sm text-gray-500">{value.publisher}</div>
          </Link>)
        }
      },
      {
        id: 'status',
        Header: 'Status',
        accessor: 'col2',
        Cell: ({ value }: any) => {
          return (<>
            <span
              className={
                !value
                  ? 'px-2 inline-flex text-xs leading-5 font-semibold rounded-full bg-green-100 text-green-800'
                  : 'px-2 inline-flex text-xs leading-5 font-semibold rounded-full bg-red-100 text-red-800'
              }
            >
              {!value ? 'Active' : 'Purgatory'}
            </span>
          </>)
        }
      },
      {
        id: 'publisher',
        Header: 'Publisher',
        accessor: 'col3',
      },
      {
        id: 'your-network',
        Header: 'Your Network',
        accessor: 'col4',
        highlight: true,
      },
      {
        id: 'times-consumed',
        Header: 'Times Consumed',
        accessor: 'col5',
      }
    ],
    []
  );

  return (
    <Layout title="Ocean Market">
      {loading && (!assets || !assets.length) ? (
        <div className="flex flex-col max-w-7xl  px-8 mt-8  mx-auto ">
          <Loader />{' '}
        </div>
      ) : (
        <Table columns={columns} data={data} />
      )}
    </Layout>
  )
}

export default Ocean
