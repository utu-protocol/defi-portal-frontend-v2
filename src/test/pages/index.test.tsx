import React from 'react'
import { render } from '../testUtils'
import { Home } from '../../pages/index'

describe('Home page', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<Home />, {})
    expect(asFragment()).toMatchSnapshot()
  })

  it('clicking button triggers alert', () => {
    const { getByText } = render(<Home />, {})
    window.alert = jest.fn()
    const title = getByText('Open main menu')
    expect(title).toBeTruthy()
  })
})
