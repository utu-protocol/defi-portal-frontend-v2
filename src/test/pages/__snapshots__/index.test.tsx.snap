// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-primary h-80"
  >
    <nav
      class="bg-transparent border-b-2 border-white border-opacity-10"
    >
      <div
        class="max-w-7xl mx-auto px-2 sm:px-6 lg:px-8"
      >
        <div
          class="relative flex items-center justify-between h-16"
        >
          <div
            class="absolute inset-y-0 left-0 flex items-center sm:hidden"
          >
            <button
              aria-expanded="false"
              class="inline-flex items-center justify-center p-2 rounded-md text-gray-400 hover:text-white hover:bg-gray-700 focus:outline-none focus:ring-2 focus:ring-inset focus:ring-white"
              id="headlessui-disclosure-button-1"
              type="button"
            >
              <span
                class="sr-only"
              >
                Open main menu
              </span>
              <svg
                aria-hidden="true"
                class="block h-6 w-6"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4 6h16M4 12h16M4 18h16"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
          </div>
          <div
            class="flex-1 flex items-center justify-center sm:items-stretch sm:justify-start"
          >
            <div
              class="flex-shrink-0 flex items-center"
            >
              <img
                alt="Workflow"
                class="block lg:hidden h-8 w-auto"
                src="https://tailwindui.com/img/logos/workflow-mark-indigo-500.svg"
              />
              <img
                alt="Workflow"
                class="hidden lg:block h-8 w-auto"
                src="https://tailwindui.com/img/logos/workflow-logo-indigo-500-mark-white-text.svg"
              />
            </div>
            <div
              class="hidden sm:block sm:ml-6"
            >
              <div
                class="flex space-x-4"
              >
                <a
                  href="/"
                >
                  Overview
                </a>
                <a
                  href="/defi"
                >
                  DeFi
                </a>
                <a
                  href="/ocean"
                >
                  Ocean
                </a>
              </div>
            </div>
          </div>
          <div
            class="absolute inset-y-0 right-0 flex items-center pr-2 sm:static sm:inset-auto sm:ml-6 sm:pr-0"
          >
            <button
              class="w-full border-white border-2 hover:text-white space-x-3 flex-row flex justify-between  rounded-md shadow shadow-sm pr-3 py-2 bg-transparent text-white font-mediumnhover:bg-white hover:bg-white hover:text-primary focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500 sm:ml-3 sm:w-auto sm:text-sm"
              type="button"
            >
              <p>
                <svg
                  aria-hidden="true"
                  class="h-6 w-6 ml-3"
                  fill="none"
                  stroke="currentColor"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M11 16l-4-4m0 0l4-4m-4 4h14m-5 4v1a3 3 0 01-3 3H6a3 3 0 01-3-3V7a3 3 0 013-3h7a3 3 0 013 3v1"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
              </p>
              <p
                class="mt-0.5"
              >
                CONNECT
              </p>
            </button>
          </div>
        </div>
      </div>
    </nav>
    <h1
      class="max-w-7xl px-8 py-12  mx-auto text-3xl font-bold text-white"
    >
      Ocean Market
    </h1>
    <div
      class="max-w-7xl mx-auto -mt-8"
    >
      <div
        class="flex flex-col max-w-7xl  px-8 py-8  mx-auto "
      >
        <div
          class="-my-2 overflow-x-auto sm:-mx-6 lg:-mx-8"
        >
          <div
            class="py-2 align-middle inline-block min-w-full sm:px-6 lg:px-8"
          >
            <div
              class="shadow overflow-hidden border-b border-gray-200 sm:rounded-lg"
            >
              <table
                class="min-w-full divide-y divide-gray-200"
              >
                <thead
                  class="bg-gray-50"
                >
                  <tr>
                    <th
                      class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                      scope="col"
                    >
                      Name
                    </th>
                    <th
                      class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                      scope="col"
                    >
                      Status
                    </th>
                    <th
                      class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                      scope="col"
                    >
                      Publisher
                    </th>
                    <th
                      class="bg-yellow-100 border-b border-yellow-200 px-6 py-3 text-left text-xs font-medium text-gray-700 uppercase tracking-wider"
                      scope="col"
                    >
                      Your Network
                    </th>
                    <th
                      class="px-6 py-3 text-center text-xs font-medium text-gray-500 uppercase tracking-wider"
                      scope="col"
                    >
                                            Total Consumption
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="bg-white divide-y divide-gray-200"
                >
                  <tr>
                    <td
                      class="px-2 py-4 whitespace-nowrap"
                    >
                      <div
                        class="flex items-center"
                        href="/ocean"
                      >
                        <div
                          class="ml-4"
                        >
                          <div
                            class="text-sm font-medium text-gray-900"
                          >
                            ðŸ‘‹ Data Union.app - Image & Annotation Vault :e
                          </div>
                          <div
                            class="text-sm text-gray-500"
                          >
                            QUICRA-0
                          </div>
                        </div>
                      </div>
                    </td>
                    <td
                      class="px-6 py-4 whitespace-nowrap"
                    >
                      <span
                        class="px-2 inline-flex text-xs leading-5 font-semibold rounded-full bg-green-100 text-green-800"
                      >
                        Active
                      </span>
                    </td>
                    <td
                      class="px-6 py-4 whitespace-nowrap"
                    >
                      <div
                        class="text-sm text-gray-900"
                      >
                         DataUnion.app
                      </div>
                      <div
                        class="text-sm text-gray-500"
                      >
                        Optimization
                      </div>
                    </td>
                    <td
                      class="bg-yellow-50 border-b border-yellow-200 border-b-yellow-200 px-6 py-4 whitespace-nowrap text-sm text-gray-500"
                    >
                      9 Consumed data assets 
                      <br />
                       3 Provided liquidity
                    </td>
                    <td
                      class="px-6 py-4 whitespace-nowrap text-left text-sm font-medium"
                    >
                      <a
                        class="font-light hover:text-indigo-900"
                        href="#"
                      >
                        302,592,120 OCEAN
                      </a>
                    </td>
                  </tr>
                  <tr>
                    <td
                      class="px-2 py-4 whitespace-nowrap"
                    >
                      <div
                        class="flex items-center"
                        href="/ocean"
                      >
                        <div
                          class="ml-4"
                        >
                          <div
                            class="text-sm font-medium text-gray-900"
                          >
                            Product Pages of 1'044'709 Products on Amazon.com...
                          </div>
                          <div
                            class="text-sm text-gray-500"
                          >
                            LUMSTA-42
                          </div>
                        </div>
                      </div>
                    </td>
                    <td
                      class="px-6 py-4 whitespace-nowrap"
                    >
                      <span
                        class="px-2 inline-flex text-xs leading-5 font-semibold rounded-full bg-red-100 text-red-800"
                      >
                        Purgatory
                      </span>
                    </td>
                    <td
                      class="px-6 py-4 whitespace-nowrap"
                    >
                      <div
                        class="text-sm text-gray-900"
                      >
                        Regional Paradigm Technician
                      </div>
                      <div
                        class="text-sm text-gray-500"
                      >
                        Optimization
                      </div>
                    </td>
                    <td
                      class="bg-yellow-50 border-b border-yellow-200 border-b-yellow-200 px-6 py-4 whitespace-nowrap text-sm text-gray-500"
                    >
                      0 Consumed data assets 
                      <br />
                       0 Provided liquidity
                    </td>
                    <td
                      class="px-6 py-4 whitespace-nowrap text-left text-sm font-medium"
                    >
                      <a
                        class="font-light hover:text-indigo-900"
                        href="#"
                      >
                        154,787,432 OCEAN
                      </a>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
